kubernetes/websocket/wsclient.c

```diff
@@ -0,0 +1,331 @@
+        if (wsc->data_to_send_len > 0 && wsc->data_to_send) {
+            int m = lws_write(wsi, wsc->data_to_send + LWS_PRE, wsc->data_to_send_len, LWS_WRITE_BINARY);
+            if (m < wsc->data_to_send_len) {
+                lwsl_err("%s: Sending message failed: %d\n", __func__, m);
+                return -1;
+            }
+            lwsl_user("%s: Succeed to send message: %d\n", __func__, m);
+            if (wsc->data_to_send) {
+                free(wsc->data_to_send);
+                wsc->data_to_send = NULL;
+                wsc->data_to_send_len = 0;
+            }
+        }
+        break;
+
+    case LWS_CALLBACK_CLIENT_CLOSED:
+        //goto do_retry;
+        g_interrupted = true;
+        break;
+    default:
+        break;
+    }
+
+    return lws_callback_http_dummy(wsi, reason, user, in, len);
+
+  do_retry:
+    /*
+     * retry the connection to keep it nailed up
+     *
+     * For this example, we try to conceal any problem for one set of
+     * backoff retries and then exit the app.
+     *
+     * If you set retry.conceal_count to be larger than the number of
+     * elements in the backoff table, it will never give up and keep
+     * retrying at the last backoff delay plus the random jitter amount.
+     */
+    if (lws_retry_sul_schedule_retry_wsi(wsi, &wsc->sul, connect_client, &wsc->retry_count)) {
+        lwsl_err("%s: wsclient connection attempts exhausted.\n", __func__);
+        g_interrupted = true;
+    }
+
+    return 0;
+}
+
+static const struct lws_protocols protocols[] = {
+    {"wsclient", callback_wsclient, 0, 0,},
+    {NULL, NULL, 0, 0}
+};
+
+static void read_from_stdin(wsclient_t * wsc)
+{
+    char wsc_attach_stdin_buffer[WSC_ATTACH_STDIN_BUFFER_SIZE];
+    memset(wsc_attach_stdin_buffer, 0, sizeof(wsc_attach_stdin_buffer));
+
+    int flag, newflag;
+    flag = fcntl(STDIN_FILENO, F_GETFL);
+    newflag = flag | O_NONBLOCK;
+    fcntl(STDIN_FILENO, F_SETFL, newflag);
+    fgets(wsc_attach_stdin_buffer, sizeof(wsc_attach_stdin_buffer) - 1, stdin);
+
+    if (wsc_attach_stdin_buffer && strlen(wsc_attach_stdin_buffer) > 0) {
+        wsc->data_to_send_len = strlen(wsc_attach_stdin_buffer) + 1 /* TTY_STDIN_NUMBER */ ;
+        wsc->data_to_send = (char *) calloc(LWS_PRE + 1 /* TTY_STDIN_NUMBER */  + wsc->data_to_send_len + 1, 1);
+        wsc->data_to_send[LWS_PRE] = TTY_STDIN_NUMBER;
+        strncpy(wsc->data_to_send + LWS_PRE + 1, wsc_attach_stdin_buffer, strlen(wsc_attach_stdin_buffer));
+    }
+
+    fcntl(STDIN_FILENO, F_SETFL, flag);
+}
+
+int wsclient_run(wsclient_t * wsc, wsc_mode_t mode)
+{
+    int n = 0;
+    lws_set_log_level(wsc->log_mask, NULL);
+
+    struct lws_context_creation_info info;
+    memset(&info, 0, sizeof(struct lws_context_creation_info));
+
+    info.options = LWS_SERVER_OPTION_DO_SSL_GLOBAL_INIT;
+    info.port = CONTEXT_PORT_NO_LISTEN; /* we do not run any server */
+    info.protocols = protocols;
+    info.fd_limit_per_thread = 1 + 1 + 1;
```

Comment:
What's the purpose of this expression vs setting the limit to just 3?

Reply:
This code comes from the sample of libwebsokets and it has below explanation for users :

```c
        /*
         * since we know this lws context is only ever going to be used with
         * one client wsis / fds / sockets at a time, let lws know it doesn't
         * have to use the default allocations for fd tables up to ulimit -n.
         * It will just allocate for 1 internal and 1 (+ 1 http2 nwsi) that we
         * will use.
         */
        info.fd_limit_per_thread = 1 + 1 + 1;
```